File: frontend/src/utils/frameSecrets.ts
import type { FrameType } from '../types'

/**
 * The deploy baseline (`frame.last_successful_deploy`) holds no secrets: the
 * backend stores an HMAC fingerprint per secret leaf instead
 * (`backend/app/utils/frame_secrets.py`) and serves the frame row's *current*
 * fingerprints beside it as `frame.secret_fingerprints`. The browser already
 * holds the current secrets from its per-frame GET, so it can fill a secret
 * back into the baseline wherever the two fingerprints agree — that secret is
 * unchanged since the deploy and compares equal — and leave it out where they
 * differ, so a rotated secret shows as "changed since deploy". Neither the
 * broadcast nor the stored snapshot ever carries the value itself.
 */

// Mirrors TOP_LEVEL_SECRET_KEYS + NESTED_SECRET_PATHS on the backend; "*"
// walks a list. A path the backend fingerprints but this list lacks makes
// that leaf "changed since deploy" forever (the baseline has no value, the
// row does), which is how "Network settings" stayed pending after every
// fast deploy once the Wi-Fi passwords became secrets — the shared-spa
// `frame-secrets-mirror` test reads the backend list and fails on drift.
export const SECRET_PATHS: readonly (readonly string[])[] = [
  ['ssh_pass'],
  ['server_api_key'],
  ['frame_access_key'],
  ['https_proxy', 'certs', 'server_key'],
  ['agent', 'agentSharedSecret'],
  ['frame_admin_auth', 'pass'],
  ['mountpoints', 'items', '*', 'password'],
  ['network', 'wifiPassword'],
  ['network', 'wifiHotspotPassword'],
]

const FINGERPRINTS_KEY = 'secret_fingerprints'

type Leaf = { container: Record<string, unknown>; key: string; dotted: string }

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function* walk(value: unknown, path: readonly string[], prefix = ''): Generator<Leaf> {
  const [head, ...rest] = path
  if (head === undefined) {
    return
  }
  if (head === '*') {
    if (!Array.isArray(value)) {
      return
    }
    for (let index = 0; index < value.length; index++) {
      yield* walk(value[index], rest, `${prefix}${index}.`)
    }
    return
  }
  if (!isRecord(value) || !(head in value)) {
    return
  }
  if (rest.length === 0) {
    yield { container: value, key: head, dotted: `${prefix}${head}` }
    return
  }
  yield* walk(value[head], rest, `${prefix}${head}.`)
}

function fingerprintsOf(value: unknown): Record<string, string> {
  const fingerprints = isRecord(value) ? value[FINGERPRINTS_KEY] : undefined
  return isRecord(fingerprints) ? (fingerprints as Record<string, string>) : {}
}

/** Deep-copies only the containers a secret can live in, so filling a leaf
 * in never mutates the frame store's snapshot object. */
function copyTouched(snapshot: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = { ...snapshot }
  for (const [container] of SECRET_PATHS) {
    if (container === undefined) {
      continue
    }
    const value = result[container]
    if (value !== null && typeof value === 'object') {
      result[container] = JSON.parse(JSON.stringify(value))
    }
  }
  return result
}

function setDotted(target: Record<string, unknown>, dotted: string, key: string, value: unknown): void {
  let cursor: unknown = target
  for (const part of dotted.split('.').slice(0, -1)) {
    if (Array.isArray(cursor)) {
      cursor = cursor[Number(part)]
    } else if (isRecord(cursor)) {
      cursor = cursor[part]
    } else {
      return
    }
  }
  if (isRecord(cursor)) {
    cursor[key] = value
  }
}

/**
 * The frame row after an `update_frame` broadcast: `incoming` over
 * `existing`, except that a secret leaf the broadcast left out of a block it
 * did send (the backend strips them — websocket_frame_payload) keeps the
 * value the browser already holds. Without this a broadcast that carried
 * `agent` would have wiped the shared secret from the form; without the
 * block at all the workspace never learnt `agent.agentVersion` after a
 * Remote upgrade until a page reload.
 */
export function mergeBroadcastFrame<T extends object>(existing: T | null | undefined, incoming: Partial<T>): T {
  const result: Record<string, unknown> = { ...(existing ?? {}), ...incoming }
  for (const path of SECRET_PATHS) {
    const container = path[0]
    if (container === undefined || path.length < 2 || !(container in incoming)) {
      continue
    }
    if (result[container] !== null && typeof result[container] === 'object') {
      result[container] = JSON.parse(JSON.stringify(result[container]))
    }
    for (const { container: from, key, dotted } of walk(existing, path)) {
      const value = from[key]
      if (value === null || value === undefined || value === '') {
        continue
      }
      let target: unknown = result
      for (const part of dotted.split('.').slice(0, -1)) {
        target = Array.isArray(target) ? target[Number(part)] : isRecord(target) ? target[part] : undefined
      }
      if (isRecord(target) && !(key in target)) {
        target[key] = value
      }
    }
  }
  return result as T
}

/**
 * `snapshot` with every secret whose stored fingerprint matches the frame's
 * current fingerprint filled in from `current`. Snapshots without
 * fingerprints (none, or a backend that predates them) come back unchanged.
 */
export function restoreDeployedSecrets<T extends Record<string, unknown>>(
  snapshot: T | null | undefined,
  current: Partial<FrameType> | null | undefined
): T | null | undefined {
  if (!isRecord(snapshot) || !(FINGERPRINTS_KEY in snapshot)) {
    return snapshot
  }
  const stored = fingerprintsOf(snapshot)
  const live = fingerprintsOf(current)
  const result = copyTouched(snapshot)
  delete result[FINGERPRINTS_KEY]
  for (const path of SECRET_PATHS) {
    for (const { container, key, dotted } of walk(current, path)) {
      const value = container[key]
      if (value === null || value === undefined || value === '') {
        continue
      }
      if (!stored[dotted] || stored[dotted] !== live[dotted]) {
        continue
      }
      setDotted(result, dotted, key, value)
    }
  }
  return result as T
}

File: backend/app/utils/frame_secrets.py
"""Secrets that live on a Frame row and must not ride along with every copy
of it.

``Frame.to_dict()`` is the wire shape for the authenticated API, but the same
dict is also broadcast on every ``update_frame`` websocket event and stored
verbatim as the ``last_successful_deploy`` baseline. Two helpers keep the
secrets out of those two places:

- ``websocket_frame_payload`` for the broadcast. The frontend merges each
  ``update_frame`` payload *shallowly* into its frame object
  (``{...state[id], ...frame}``), so a key that is present replaces the
  client's copy wholesale and a key that is absent leaves it untouched.
  Top-level secrets are therefore popped, and so are the whole objects that
  carry a nested secret (``https_proxy``, ``agent``, ``frame_admin_auth``,
  ``mountpoints``): sending them with the secret leaf removed would blank the
  client's copy of that leaf, which the settings form then saves back as
  empty. Popping the container keeps the client's last GET intact.

- ``deploy_snapshot`` / ``restore_snapshot_secrets`` for the stored baseline.
  The snapshot is a diff baseline ("what changed since the last deploy"), so
  it cannot simply lose the secrets: every consumer would read a missing key
  as a change. The stored copy keeps an HMAC fingerprint per secret instead,
  and ``restore_snapshot_secrets`` fills the *current* value back in wherever
  the fingerprint still matches — so an unchanged secret compares equal and a
  rotated one shows as changed, without the old value ever being kept.
"""
from __future__ import annotations

import copy
import hashlib
import hmac
from typing import Any, Iterator

from app.config import config

# Top-level Frame.to_dict() keys that are secrets outright.
TOP_LEVEL_SECRET_KEYS = ("ssh_pass", "server_api_key", "frame_access_key")

# Secret leaves inside nested objects; "*" iterates a list.
NESTED_SECRET_PATHS = (
    ("https_proxy", "certs", "server_key"),
    ("agent", "agentSharedSecret"),
    ("frame_admin_auth", "pass"),
    ("mountpoints", "items", "*", "password"),
)

# The top-level objects a nested secret lives in. Popped whole from websocket
# payloads (see the module docstring for why).
SECRET_CONTAINER_KEYS = tuple(dict.fromkeys(path[0] for path in NESTED_SECRET_PATHS))

FINGERPRINTS_KEY = "secret_fingerprints"


def _walk(value: Any, path: tuple[str, ...], prefix: str = "") -> Iterator[tuple[Any, str, str]]:
    """Yield ``(container, key, dotted_path)`` for every leaf that ``path``
    addresses inside ``value``; skips whatever is missing or not a dict/list."""
    head, rest = path[0], path[1:]
    if head == "*":
        if not isinstance(value, list):
            return
        for index, item in enumerate(value):
            yield from _walk(item, rest, f"{prefix}{index}.")
        return
    if not isinstance(value, dict) or head not in value:
        return
    if not rest:
        yield value, head, f"{prefix}{head}"
        return
    yield from _walk(value[head], rest, f"{prefix}{head}.")


def _secret_paths() -> tuple[tuple[str, ...], ...]:
    return tuple((key,) for key in TOP_LEVEL_SECRET_KEYS) + NESTED_SECRET_PATHS


def _copy_touched(frame_dict: dict) -> dict:
    """A copy of ``frame_dict`` that shares nothing with the secret-bearing
    parts, so removing leaves cannot mutate the ORM's JSON columns."""
    result = dict(frame_dict)
    for key in SECRET_CONTAINER_KEYS:
        if key in result:
            result[key] = copy.deepcopy(result[key])
    return result


def redact_frame_secrets(frame_dict: dict) -> dict:
    """``frame_dict`` with every set secret leaf removed (containers kept;
    an empty value is not a secret and stays so shapes round-trip)."""
    result = _copy_touched(frame_dict)
    for path in _secret_paths():
        for container, key, _ in list(_walk(result, path)):
            if container.get(key) not in (None, ""):
                container.pop(key, None)
    return result


def websocket_frame_payload(frame_dict: dict) -> dict:
    """The ``update_frame`` broadcast shape: secrets and the objects that
    carry them are omitted, never sent empty (shallow client merge)."""
    payload = dict(frame_dict)
    for key in TOP_LEVEL_SECRET_KEYS + SECRET_CONTAINER_KEYS:
        payload.pop(key, None)
    # ``last_successful_deploy`` is left as ``to_dict()`` serialized it: the
    # frontend diffs it key by key against its own copy of the frame, so a
    # stripped baseline would report every secret as "changed since deploy"
    # after each deploy. What it carries is only what already matches the
    # current row (restore_snapshot_secrets), never a superseded value.
    return payload


def _fingerprint_keys() -> list[str]:
    return [key for key in (config.SECRET_KEY, *config.PREVIOUS_SECRET_KEYS) if key]


def _fingerprint(value: Any, key: str) -> str:
    return hmac.new(key.encode("utf-8"), str(value).encode("utf-8"), hashlib.sha256).hexdigest()


def frame_secret_fingerprints(frame_dict: dict) -> dict[str, str]:
    """``{dotted_path: hmac}`` for every non-empty secret in ``frame_dict``."""
    keys = _fingerprint_keys()
    if not keys:
        return {}
    fingerprints: dict[str, str] = {}
    for path in _secret_paths():
        for container, key, dotted in _walk(frame_dict, path):
            value = container.get(key)
            if value not in (None, ""):
                fingerprints[dotted] = _fingerprint(value, keys[0])
    return fingerprints


def deploy_snapshot(frame_dict: dict) -> dict:
    """What ``last_successful_deploy`` stores: the frame minus its secrets,
    plus their fingerprints so the baseline diff still works."""
    snapshot = redact_frame_secrets(frame_dict)
    snapshot.pop("last_successful_deploy", None)
    snapshot.pop("last_successful_deploy_at", None)
    fingerprints = frame_secret_fingerprints(frame_dict)
    if fingerprints:
        snapshot[FINGERPRINTS_KEY] = fingerprints
    return snapshot


def restore_snapshot_secrets(snapshot: Any, current: dict) -> Any:
    """The stored snapshot as consumers expect it: each secret whose
    fingerprint matches the value in ``current`` is filled back in from
    ``current``; the rest stay absent. Snapshots written before fingerprints
    existed are returned unchanged."""
    if not isinstance(snapshot, dict) or FINGERPRINTS_KEY not in snapshot:
        return snapshot
    fingerprints = snapshot.get(FINGERPRINTS_KEY) or {}
    result = _copy_touched(snapshot)
    result.pop(FINGERPRINTS_KEY, None)
    keys = _fingerprint_keys()
    for path in _secret_paths():
        for container, key, dotted in _walk(current, path):
            value = container.get(key)
            stored = fingerprints.get(dotted)
            if value in (None, "") or not stored:
                continue
            if not any(hmac.compare_digest(stored, _fingerprint(value, k)) for k in keys):
                continue
            target = result
            for part in dotted.split(".")[:-1]:
                if isinstance(target, list):
                    index = int(part)
                    target = target[index] if index < len(target) else None
                else:
                    target = target.get(part) if isinstance(target, dict) else None
                if target is None:
                    break
            if isinstance(target, dict):
                target[key] = value
    return result


def deployed_frame_snapshot(frame: Any) -> Any:
    """``frame.last_successful_deploy`` with matching secrets restored from
    the frame's current columns. Works on ORM rows and on the SimpleNamespace
    stand-ins the deploy tests use."""
    snapshot = getattr(frame, "last_successful_deploy", None)
    if not isinstance(snapshot, dict):
        return snapshot
    current = {
        key: getattr(frame, key, None)
        for key in TOP_LEVEL_SECRET_KEYS + SECRET_CONTAINER_KEYS
    }
    return restore_snapshot_secrets(snapshot, current)
